### PR TITLE
Versioned serialization

### PR DESCRIPTION
**Pull request by tombujok**
Prepares the ground for the versioned serialization implemented in the EE.

### CODE REVIEW

**Comment by tombujok** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/core/Cluster.java [hunk 1]
```diff
@@ -53,7 +54,7 @@
      * Removes the specified MembershipListener.
      * <p/>
      * If the same MembershipListener is registered multiple times, it needs to be removed multiple times.
-     *
+     * <p>
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/core/Cluster.java]: You can disable this in Idea :)

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java [hunk 2]
```diff
@@ -63,5 +64,7 @@
 
     SerializationServiceBuilder setInitialOutputBufferSize(int initialOutputBufferSize);
 
+    SerializationServiceBuilder setClusterService(ClusterService clusterService);
```
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java]: yeah, I know, it's not perfect. I need a clusterVersion on the write/read call.
we can also change the signatures, but it's used sometimes only

File: hazelcast/src/main/java/com/hazelcast/core/Cluster.java (hunk 1 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/core/Cluster.java]: yeah, I have to finally do it actually :)

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java (hunk 2 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java]: Aha, and some serializers may actually not need it at all, so that was the reason why I didn't change the `StreamSerializer` API.

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java]: To reduce coupling, the Cluster version (String?) could be used instead of ClusterService.

Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java]: The thing is that the clusterVersion changes in runtime so I need the `newest` version while serialising

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializerOld.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
+/*
```
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializerOld.java]: will be deleted from the PR

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/Versioned.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/Versioned.java]: will be in impl package.

File: hazelcast/src/main/java/com/hazelcast/core/Cluster.java [hunk 5]
```diff
@@ -216,4 +217,6 @@
      * @since 3.6
      */
     void shutdown(TransactionOptions transactionOptions);
+
+    Version getClusterVersion();
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/core/Cluster.java]: Is `Version` a public api?

Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/core/Cluster.java]: yep. it will be used to do an upgrade by the user.

Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/core/Cluster.java]: ok, nicely spot! I'll double check those

Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/core/Cluster.java]: Fixed

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java (hunk 2 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java]: I understand this. But I feel hesitant introducing a dependency on a big object like ClusterService for the sake of obtaining a version. Of course it is the easiest thing to do in the short term, but it increases coupling between the 2 services.

Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java]: What about registering the `SerializationService` as a cluster version listener? (not part of this PR); it would require some listener support code on the `ClusterStateManager` side, but should reduce coupling.

Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java]: Sure, I understand this.
I need the version though. Two options I see is to change the service's APIe to pass the version on the serialization call (it seems ugly). 
The other option is to have some coupling, we can introduce a new interface (a sub-interface of the ClusterService).
A listener sounds good too - don't know how feasible that is.

Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/SerializationServiceBuilder.java]: Fixed by introducing a versionListener in the EE.

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/ClusterVersionListener.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * Listeners interested on acting upon cluster version update should implement this interface.
+ * TODO: only for services? (like ClusterStateListener?)
+ * TODO: can clients listen for cluster version change?
+ *
+ * @see Cluster#getClusterVersion()
+ * @see Cluster#changeClusterVersion(Version)
+ */
+public interface ClusterVersionListener {
+
```
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/ClusterVersionListener.java]: This will come from Vassiliss' PR

**Comment by pveentjer** [comment]
Looks good

File: hazelcast/src/main/java/com/hazelcast/version/Version.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.DataSerializable;
+
+import java.io.IOException;
+
+/**
+ * Determines Hazelcast version.
+ * Immutable.
+ */
+public final class Version implements DataSerializable {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/version/Version.java]: also constructors should not be public? given the class itself is part of public API

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/VersionedObjectDataOutput.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import com.hazelcast.version.Version;
+
+import java.io.OutputStream;
+
+/**
+ * Base class for ObjectDataOutput that is VersionAware and allows mutating the version
+ */
+abstract class VersionedObjectDataOutput extends OutputStream implements ObjectDataOutput, VersionAware {
+
+    protected Version version = Version.UNKNOWN;
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/VersionedObjectDataOutput.java]: what's the semantic of this? under what circumstances it can return `UKNOWN` version?

File: hazelcast/src/main/java/com/hazelcast/version/Version.java (hunk 7 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/version/Version.java]: Class removed, since provided in another PR.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/VersionedObjectDataInput.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+abstract class VersionedObjectDataInput extends InputStream implements ObjectDataInput, VersionAware {
+
+    protected Version version = Version.UNKNOWN;
+
+    public void setVersion(Version version) {
+        this.version = version;
+    }
+
+    @Override
+    public Version getVersion() {
```
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/VersionedObjectDataInput.java]: Fixed. More doc will be in EE that actually uses it.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/VersionedObjectDataOutput.java (hunk 8 above)
Comment (tombujok) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/VersionedObjectDataOutput.java]: Fixed. More doc will be in EE that actually uses it.

**Comment by tombujok** [comment]
run-lab-run
